### DISCUSSION TITLE
Why is there an upper limit for Tpetra::CrsMatrix constructor?

### DISCUSSION THREAD

**Discussion by abodado** [trilinos/Trilinos#13562] [Q&A] [answered]
@tpetra, @csiefer2 
Just curious as to why there is an upper limit on the max number of entries per row on the Tpetra::CrsMatrix constructor. This is a different design compared to Epetra, where in Epetra, if trying to insert more than the "suggested" max entries in a given row, more memory would be allocated. Is there any documentation as to why this design change was made? What are the benefits of this design change? I see that there is another constructor where I can specify an array that contains the max number of entries per row. I suppose I will have to go that route but it does add complexity to my code, so I was wondering if there are any other work-arounds or if Tpetra will at some point allow going over the max entry per row with the additional memory allocation hit, as occurs in Epetra.

**Comment by jhux2** [comment, marked as answer]
[18_TUG.pdf](https://github.com/user-attachments/files/17736452/18_TUG.pdf)

@abodado This was an intentional design decision to remove the "dynamic" allocation in Tpetra CRS matrix construction.  While convenient for the user, there are potentially large performance penalties, especially on GPU architectures.

[EDIT]:  See slide 10 of the attached presentation.

**Reply by abodado** [reply]
@jhux2 , very good. Thank you for that resource. I have updated my code to handle the new requirement.
